java/src/main/java/com/logicalclocks/hsfs/engine/SparkEngine.java

```diff
@@ -786,6 +790,53 @@ public <S> List<Feature> parseFeatureGroupSchema(S datasetGeneric,
     return features;
   }
 
+  public Dataset<Row> castColumnType(Dataset<Row> dataset, List<TrainingDatasetFeature> features)
+      throws FeatureStoreException {
+    for (TrainingDatasetFeature feature: features) {
+      dataset = dataset.withColumn(
+          feature.getName(),
+          dataset.col(feature.getName()).cast(convertColumnType(feature.getType()))
+      );
+    }
+    return dataset;
+  }
+
+  private static Pattern arrayPattern = Pattern.compile("^array<(.*)>$");
+
+  private DataType convertColumnType(String type) throws FeatureStoreException {
+    Matcher m = arrayPattern.matcher(type);
+    if (m.matches()) {
+      return DataTypes.createArrayType(convertColumnType(m.group(1)));
+    } else if (type.contains("struct<label:string,index:int>")) {
```

Comment:
Our struct types are more general, we support `STRUCT<NAME: TYPE, ...>`. But the Spark CSV writer does not support Array, Map, and Struct types (I just checked). Since this is only used in the context of reading CSV's that have been written by the spark client, we can ommit the Array and Struct type and move the logic that is currently in `convertBasicType` here.